bluegreen_deploy.py

```diff
@@ -65,226 +66,305 @@ def list_marathon_apps(args):
+
+
+def fetch_app_listeners(app, marathon_lb_urls):
+    haproxy_stats = fetch_combined_haproxy_stats(marathon_lb_urls)
+    return [l for l in haproxy_stats if _if_app_listener(app, l)]
+
+
+def waiting_for_listeners(new_app, old_app, listeners, haproxy_count):
+    listener_count = (len(listeners) / haproxy_count)
+    return listener_count != new_app['instances'] + old_app['instances']
+
+
+def get_deployment_target(app):
+    return int(app['labels']['HAPROXY_DEPLOYMENT_TARGET_INSTANCES'])
+
+
+def waiting_for_up_listeners(app, listeners, haproxy_count):
+    up_listeners = [l for l in listeners if l.status == 'UP']
+    up_listener_count = (len(up_listeners) / haproxy_count)
+
+    return up_listener_count < get_deployment_target(app)
+
+
+def _has_pending_requests(listener):
+    return int(listener.qcur or 0) > 0 or int(listener.scur or 0) > 0
+
+
+def select_drained_listeners(listeners):
+    draining_listeners = [l for l in listeners if l.status == 'MAINT']
+    return [l for l in draining_listeners if not _has_pending_requests(l)]
+
+
+def get_svnames_from_task(task):
+    prefix = '{}_{}_{}_{}'.format(*task['host'].split('.'))
+    for port in task['ports']:
+        yield(prefix + '_{}'.format(port))
+
+
+def get_svnames_from_tasks(tasks):
+    svnames = []
     for task in tasks:
-        if task['host'] in hostports:
-            for port in hostports[task['host']]:
-                if port in task['ports']:
-                    tasks_to_kill.add(task['id'])
-    return list(tasks_to_kill)
+        svnames += get_svnames_from_task(task)
 
+    return svnames
 
-def check_if_tasks_drained(args, app, existing_app, step_started_at):
-    time.sleep(args.step_delay)
-    url = args.marathon + "/v2/apps" + existing_app['id']
-    response = requests.get(url, auth=get_marathon_auth_params(args))
-    response.raise_for_status()
-    existing_app = response.json()['app']
 
-    url = args.marathon + "/v2/apps" + app['id']
-    response = requests.get(url, auth=get_marathon_auth_params(args))
-    response.raise_for_status()
-    app = response.json()['app']
+def find_drained_task_ids(app, listeners, haproxy_count):
+    """Return app tasks which have all haproxy listeners down and drained
+       of any pending sessions or connections
+    """
+    tasks = zip(get_svnames_from_tasks(app['tasks']), app['tasks'])
+    drained_listeners = select_drained_listeners(listeners)
+
+    drained_task_ids = []
+    for svname, task in tasks:
+        if len([l for l in drained_listeners if l.svname == svname]) == \
+          haproxy_count:
+            drained_task_ids.append(task['id'])
+
+    return drained_task_ids
+
+
+def max_wait_exceeded(args, timestamp):
+    return (time.time() - timestamp > args.max_wait)
+
+
+def check_time_and_sleep(args, timestamp):
+    if max_wait_exceeded(args, timestamp):
+        raise Exception('Max wait Time Exceeded')
+
+    return time.sleep(args.step_delay)
```

Comment:
Same thing here as https://github.com/JamesAwesome/marathon-lb/pull/1/files#r61812509: just pass the individual parameters, rather than a big fat argument blob.